src/Toml/BiMap.hs

```diff
@@ -293,3 +299,148 @@ _HashSet bimap = iso HS.toList HS.fromList >>> _Array bimap
 -- 'intSet' combinator.
 _IntSet :: BiMap IS.IntSet AnyValue
 _IntSet = iso IS.toList IS.fromList >>> _Array _Int
+
+----------------------------------------------------------------------------
+-- Generic data structures
+----------------------------------------------------------------------------
+
+-- | Class to generate a 'BiMap' from any generic data structure to 'Anything'.
+class GenBiMap a where
+  -- | Encoding the data structure
+  _encode :: a -> Maybe AnyValue
+  -- | Adding extra constraints
+  default _encode :: (Generic a, GenBiMap' (Rep a)) => a -> Maybe AnyValue
+  _encode = encode' . from
+
+  -- | Decoding the data structure
+  _decode :: AnyValue -> Maybe a
+  -- | Adding extra constraints
+  default _decode :: (Generic a, GenBiMap' (Rep a)) => AnyValue -> Maybe a
+  _decode = fmap to . decode'
+
+-- | Helper class needed to implement 'GenBiMap'.
+class GenBiMap' f where
+  encode' :: f p -> Maybe AnyValue
+  decode' :: AnyValue -> Maybe (f p)
+
+-- | Instance for for datatypes without constructors.
+instance GenBiMap' V1 where
+  encode' _ = undefined
+  decode' _ = undefined
+
+-- | Instance for constructors without arguents.
+instance GenBiMap' U1 where
+  encode' U1 = Just $ AnyValue $ Array []
+
+  decode' (AnyValue (Array [])) = Just U1
+  decode' _                     = Nothing
+
+-- | Instance for sum types.
+instance (GenBiMap' f, GenBiMap' g) => GenBiMap' (f :+: g) where
+  encode' = \case
+      (L1 x) -> encode' x >>= go False
+      (R1 x) -> encode' x >>= go True
+    where
+      go :: Bool -> AnyValue -> Maybe AnyValue
+      go b (AnyValue v) = Just $ AnyValue $ Array [Array [Bool b], Array [v]]
+
+  decode' (AnyValue (Array [Array [Bool False], Array [v]])) =
+               L1 <$> decode' (AnyValue v)
+  decode' (AnyValue (Array [Array [Bool  True], Array [v]])) =
+               R1 <$> decode' (AnyValue v)
+  decode' _ = Nothing
+
+-- | Instance for product types.
+instance (GenBiMap' f, GenBiMap' g) => GenBiMap' (f :*: g) where
```

Comment:
At this point this typeclass looks more like `ToValue/FromValue` as was discussed earlier. After looking closely at the implementation, I don't think there's big value (no pun intended) in such generic instances for `BiMap`. Because instance for `:*:` is a product of types, basically, records. And you shouldn't use `BiMap` with complicated data types anyways.